File: python3/problems/264.丑数-ii.py
#
# @lc app=leetcode.cn id=264 lang=python3
#
# [264] 丑数 II
#

# @lc code=start


import heapq


class Solution:
    def nthUglyNumber(self, n: int) -> int:
        nums = [2, 3, 5]
        set = {1}
        pq = [1]

        for i in range(1, n + 1):
            x = heapq.heappop(pq)
            if i == n:
                return x
            for num in nums:
                t = num * x
                if t not in set:
                    set.add(t)
                    heapq.heappush(pq, t)
        return -1


# @lc code=end

File: typescript/problems/264.丑数-ii.ts
/*
 * @lc app=leetcode.cn id=264 lang=typescript
 *
 * [264] 丑数 II
 */

// @lc code=start
function nthUglyNumber(n: number): number {
  const ans = new Array(n + 1).fill(1)
  for (let i2 = 1, i3 = 1, i5 = 1, i = 2; i <= n; i++) {
    let a = ans[i2] * 2,
      b = ans[i3] * 3,
      c = ans[i5] * 5
    const min = Math.min(a, b, c)
    if (min === a) i2++
    if (min === b) i3++
    if (min === c) i5++
    ans[i] = min
  }
  return ans[n]
}
// @lc code=end
